discovery/README.md

```diff
@@ -121,3 +121,42 @@ or incorrect metadata.
 The information obtained from service discovery is not considered sensitive
 security wise. Do not return secrets in metadata, anyone with access to
 the Prometheus server will be able to see them.
+
+
+## Writing an SD mechanism
+
+### The SD interface
+
+A Service Discovery (SD) mechanism has to discover targets and provide them to Prometheus. It has to send the targets down to Prometheus in the form of [`[]TargetGroup`](https://godoc.org/github.com/prometheus/prometheus/config#TargetGroup).
+
+For example take the case of [File SD](https://prometheus.io/docs/operating/configuration/#file_sd_config) with the following file as input:
+
+```
+- targets: ['10.11.150.1:7870', '10.11.150.4:7870']
+  labels:
+    job: mysql
+
+- targets: ['10.11.122.11:6001', '10.11.122.15:6002']
+  labels:
+    job: postgres
+```
+
+Here there are two `TargetGroups` defined in the file, one group with `job: mysql` and another with `job: postgres`. The grouping is implementation specific and could even be one target per group. But, one has to make sure every target group sent by an SD instance should have a `Source` which is unique across all the `TargetGroup`s of that SD instance. 
```

Comment:
I am trying to move this to the structs but it becomes complicated especially with the `LabelSet`: https://godoc.org/github.com/prometheus/common/model#LabelSet. I think this is fair enough.